code/modules/mob/living/simple_animal/hostile/hostile.dm

```diff
@@ -543,12 +554,54 @@
 /mob/living/simple_animal/hostile/proc/CheckAndAttack()
 	if(!target)
 		return FALSE
-	var/in_range = melee_reach > 1 ? target.Adjacent(targets_from) || (get_dist(src, target) <= melee_reach && (target in view(src, melee_reach))) : target.Adjacent(targets_from)
+	var/in_range = melee_reach > 1 ? target.Adjacent(targets_from) || (get_dist(src, target) <= melee_reach && (target in view(melee_reach, src))) : target.Adjacent(targets_from)
 	if(targets_from && isturf(targets_from.loc) && in_range && !incapacitated())
 		AttackingTarget()
 		return TRUE
 	return FALSE
 
+/mob/living/simple_animal/hostile/Bumped(atom/movable/AM)
+	. = ..()
+	if(!client && AIStatus == AI_ON && !attack_is_on_cooldown && CanAttack(AM))
+		TryAttack()
+
+/mob/living/simple_animal/hostile/Moved()
+	. = ..()
+	if(!client && AIStatus == AI_ON && target && !attack_is_on_cooldown)
+		TryAttack()
+
+/mob/living/simple_animal/hostile/proc/TryAttack(called_by_timer = FALSE)
```

Comment:
CheckAndAttack seems otherwise abandoned here. I believe it could and should be implemented for better range detection along with a more efficient check; the previous does not run `view` if the user does not have increased melee range to reduce calculations, the newer one I believe runs `view` regardless.

Reply:
This whole proc is basically check and attack and normally if the first part of && is false the second is not evaluated